### PR TITLE
TAUR-1137 Aggregate duplicated tweets.  A tweet is a duplicate if it …

### PR DESCRIPTION
**Pull request by oxtopus**
…has the same filtered text within the same agg_ts period.  "RT" prefix and url suffix are stripped according to current client behavior.  This prevents massive spikes over short periods of time.

re: https://jira.numenta.com/browse/TAUR-1137

Also includes some unrelated fixes due to namespace changes in https://github.com/numenta/numenta-apps/pull/5
Do not merge yet, however, until discuss with @lscheinkman and @jcasner

### CODE REVIEW

File: taurus/taurus/engine/runtime/dynamodb/dynamodb_service.py [hunk 1]
```diff
@@ -473,22 +476,82 @@ def _handleModelInferenceResults(self, body):
...
     g_log.info("Handling %d non-metric tweet item(s)", len(payload))
-    with self._metric_tweets.batch_write() as dynamodbBatchWrite:
-      for item in payload:
-        item["metric_name_tweet_uid"] = (
-          "-".join((item["metric_name"], item["tweet_uid"])))
-        data = MetricTweetsDynamoDBDefinition().Item(**item)
-        dynamodbBatchWrite.put_item(data=data._asdict(), overwrite=True)
+    for item in payload:
+
+      item["metric_name_tweet_uid"] = (
```
Comment (lscheinkman) [inline taurus/taurus/engine/runtime/dynamodb/dynamodb_service.py]: This attribute is no longer used with the new schema

File: taurus/taurus/engine/runtime/dynamodb/dynamodb_service.py [hunk 2]
```diff
@@ -473,22 +476,82 @@ def _handleModelInferenceResults(self, body):
...
+
+        try:
+          updateItemWithRetries()
+          g_log.info("Duplicate tweet.  Atomically increased copy_count of "
+                     "original tweet.  text=%s; agg_ts=%s;",
+                     tweet_text, item["agg_ts"])
+        except Exception:
+          g_log.exception("Atomic of update copy_count failed: table=%s; "
+                          "updateKey=%s; condition_expression=%s; "
+                          "expression_attribute_values=%s;",
```
Comment (shantanoo) [inline taurus/taurus/engine/runtime/dynamodb/dynamodb_service.py]: `s/;/\./`?

**Comment by oxtopus** [comment]
@lscheinkman With 4fce568 the index has changed such that the range key is metric_name + sort_key where sort_key is a number as discussed -- "142437500400001" for example.  I'm multiplying the timestamp by 1e5 and incrementing with duplicates.

**Comment by oxtopus** [comment]
Merging this.  @jcasner @lscheinkman please understand that this is a _*breaking*_ change with relation to the metric tweets table and the decision to move to production should not be taken lightly.

File: taurus/taurus/engine/runtime/dynamodb/definitions/metric_tweets_dynamodbdefinition.py [hunk 3]
```diff
@@ -45,7 +46,7 @@ def tableName(self):
   def tableCreateKwargs(self):
     return dict(
       schema=[
-        HashKey("metric_name_tweet_uid"),
```
Comment (lscheinkman) [inline taurus/taurus/engine/runtime/dynamodb/definitions/metric_tweets_dynamodbdefinition.py]: @oxtopus `metric_name_tweet_uid` is not being used and should be removed from the schema https://github.com/numenta/numenta-apps/pull/34/files#diff-5e812779471ee3b141780baf1a820886L81
and from the message handler https://github.com/numenta/numenta-apps/pull/34/files#r32277990

File: taurus/tests/integration/runtime/dynamodb/dynamodb_service_test.py [hunk 4]
```diff
@@ -286,6 +300,56 @@ def gracefulDelete():
...
+      # Assert same as original, except for copy_count, which should be 1
+
+      self.assertEqual(metricTweetItem["username"], twitterData[0]["username"])
+      self.assertEqual(metricTweetItem["tweet_uid"], twitterData[0]["tweet_uid"])
+      self.assertEqual(metricTweetItem["created_at"], twitterData[0]["created_at"])
+      self.assertEqual(metricTweetItem["agg_ts"], twitterData[0]["agg_ts"])
+      self.assertEqual(metricTweetItem["text"], twitterData[0]["text"])
+      self.assertEqual(metricTweetItem["userid"], twitterData[0]["userid"])
+      self.assertEqual(metricTweetItem["username"], twitterData[0]["username"])
+      self.assertEqual(metricTweetItem["retweet_count"], twitterData[0]["retweet_count"])
```
Comment (shantanoo) [inline taurus/tests/integration/runtime/dynamodb/dynamodb_service_test.py]: We can reduce this by using `for` look.

```
for key in ["username", "tweet_uid", ....]:
  self.assertEqual(metricTweetItem[key], twitterData[0][key])
```

File: taurus/tests/integration/runtime/dynamodb/dynamodb_service_test.py [hunk 5]
```diff
@@ -286,6 +300,56 @@ def gracefulDelete():
...
+
+      # Assert same as original, except for copy_count, which should be 1
+
+      self.assertEqual(metricTweetItem["username"], twitterData[0]["username"])
+      self.assertEqual(metricTweetItem["tweet_uid"], twitterData[0]["tweet_uid"])
+      self.assertEqual(metricTweetItem["created_at"], twitterData[0]["created_at"])
+      self.assertEqual(metricTweetItem["agg_ts"], twitterData[0]["agg_ts"])
+      self.assertEqual(metricTweetItem["text"], twitterData[0]["text"])
+      self.assertEqual(metricTweetItem["userid"], twitterData[0]["userid"])
+      self.assertEqual(metricTweetItem["username"], twitterData[0]["username"])
```
Comment (shantanoo) [inline taurus/tests/integration/runtime/dynamodb/dynamodb_service_test.py]: This is duplicate of line 339.

File: taurus/tests/integration/runtime/dynamodb/dynamodb_service_test.py [hunk 6]
```diff
@@ -286,6 +300,56 @@ def gracefulDelete():
...
+        twitterData[0]["agg_ts"]
+      )
+
+      if metricTweetItem["copy_count"] != 1:
+        time.sleep(1)
+        continue
+
+      # Assert same as original, except for copy_count, which should be 1
+
+      self.assertEqual(metricTweetItem["username"], twitterData[0]["username"])
```
Comment (shantanoo) [inline taurus/tests/integration/runtime/dynamodb/dynamodb_service_test.py]: If the sequence does not matter, can we sort these asserts on `key`?

File: nta.utils/nta/utils/message_bus_connector.py [hunk 7]
```diff
@@ -33,8 +33,12 @@
 
 import requests
 
-from nta.utils import error_handling
 from nta.utils import amqp
+from nta.utils.amqp.constants import AMQPDeliveryModes, AMQPErrorCodes
```
Comment (vitaly-krugl) [inline nta.utils/nta/utils/message_bus_connector.py]: These imports are not used

File: nta.utils/nta/utils/message_bus_connector.py [hunk 8]
```diff
@@ -64,6 +68,7 @@ class ConsumerCancelled(MessageBusConnectorError):
   retryExceptions=(
     amqp.exceptions.AmqpChannelError,
     amqp.exceptions.AmqpConnectionError,
+    amqp.exceptions.NackError,
```
Comment (vitaly-krugl) [inline nta.utils/nta/utils/message_bus_connector.py]: @oxtopus, how were you able to reproduce `NackError`?

**Pull request merged**